pkg/crypto/hash.go

```diff
@@ -0,0 +1,83 @@
+package crypto
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"hash"
+	"io"
+	"strings"
+
+	"golang.org/x/crypto/sha3"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	// DefaultHasher is the default implementation for hashing things
+	defaultHasher = basicHasher{sha3.New256()}
+)
+
+// Hash is a convenient function calling the default hasher
+func Hash(data ...interface{}) ([]byte, error) {
+	return defaultHasher.Hash(data...)
+}
+
+// HashToString is a convenient function calling the default hasher and encoding the result as hex string
+func HashToString(data ...interface{}) (string, error) {
+	hash, err := defaultHasher.Hash(data...)
+	if err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(hash), nil
+}
+
+// Hasher provides a method for hashing arbitary data types
+type Hasher interface {
+	Hash(data ...interface{}) ([]byte, error)
+}
+
+type basicHasher struct {
+	hash hash.Hash
+}
+
+func (h basicHasher) Hash(args ...interface{}) ([]byte, error) {
+	h.hash.Reset()
+
+	for _, data := range args {
+		var (
+			reader io.Reader
+			wg     = &errgroup.Group{}
+		)
+
+		// setup reader for the data
+		switch d := data.(type) {
+		case io.Reader:
```

Comment:
why not check for the `Stringer` and `json.Marshaller` interfaces, the default could stay fail with an error right? If we check those interfaces it will reduce the error cases significantly

Reply:
`Stringer` makes sense for sure, but not everything that is json marshalable implements the json.Marshaller interface. Things like `int` or `map[string]interface{}` wouldn't be covered. Even `[]string` wouldn't work.

Reply:
The only real error cases are when we try to pass in data which contains cyclic references, or maps that have non-strings as keys. I think the simplest solution would be to mention these cases in the docs, or even better use yaml and only mention the first case as a warning.